apps/docusaurus/docs/move/move-on-aptos/move-security-guidelines.md

```diff
@@ -589,3 +647,104 @@ public fun main(user: &signer) {
+            mint_to_user(user);
+        }
+    }
+}
+```
+
+In this example, the `play` function is `public`, allowing it to be composed with other modules. A malicious user can invoke this function and then check if they have won. If they have not won, they can abort the transaction and try again.
+
+```move
+module attacker::exploit {
+    public entry fun exploit(attacker: &signer) {
+        @user::lottery::play(attacker);
+        assert!(exists<@user::lottery::WIN>(address_of(attacker)));
+    }
+}
+```
+
+To resolve the possible issue, is sufficient to set the visibility of all functions that invoke the randomness API, either directly or indirectly, to `entry` rather than `public` or `public entry`.
+
+#### Example Secure Code
+
+```move
+module user::lottery {
+    fun mint_to_user(user: &signer){
+        move_to(address_of(user),Object_for_the_winner{});
+    }
+
+    #[randomness]
+    entry fun play(user: &signer) {
+        let random_value = aptos_framework::randomness::u64_range(0, 100);
+        if (random_value == 42) {
+            mint_to_user(user);
+        }
+    }
+}
+```
+
+> At Aptos, We are always security-first. During compilation, we ensure that no randomness API is invoked from a public function. However, we still allow users to make this choice by adding the attribute `#[lint::allow_unsafe_randomness]` to the public function.
+
+### Randomness - undergasing
+
+When different code paths in a function consume different amounts of gas, an attacker can manipulate the gas limit to bias the outcome.
+
+If the "lose" path consumes more gas than the "win" path, an attacker can set the gas limit such that the "lose" path will run out of gas. The transaction will be aborted and the user will never lose.
+
+#### Example Vulnerable code
+
+```
+module user::lottery {
+    fun win(user: &signer) {
+        [...]
+    }
+
+    fun lose(user: &signer) {
+        [...]
+    }
+    #[randomness]
+    entry fun play(user: &signer) {
+        let random_value = aptos_framework::randomness::u64_range(0, 100);
+        if (random_value == 42) {
+            win(user);
+        } else {
+            lose(user);
+        }
+    }
+}
+```
+
+In this lottery-example, `win` and `lose` consume different amounts of gas.
+If `lose` consumes more gas than `win`, an attacker can set the gas limit that is sufficient for `win` but not for `lose`. This forces the transaction to abort when the "lose" path is taken, ensuring that the user never submits a valid transaction for the "lose" path.
+
+#### Example Secure Code
+
+There are different ways to secure the code:
+
+1. Make sure better outcomes use more gas.
+2. Allows only admin addresses to invoke randomness API.
+3. Make entry functions work regardless of random outcomes. Avoid immediate actions based on randomness for consistent gas use.
+
+> We will be providing more functionality in the future, to allow for more complex code to be able to be safe against undergasing attacks.
```

Comment:
should we add a link to the complete randomness documentation?

Reply:
+1 please add appropriate randomness docs link here

Reply:
Added at the beginning of the randomness section